server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -288,13 +296,132 @@ private static class FromSingle extends AutoDateHistogramAggregator {
+            long bestDuration = (high - low) / targetBuckets;
+            long interval = 0;
+            roundingInfosLoop: do {
+                RoundingInfo curRoundingInfo = roundingInfos[roundingIdx];
+                for (int curInnerInterval: curRoundingInfo.innerIntervals) {
+                    if (bestDuration <= curInnerInterval * curRoundingInfo.roughEstimateDurationMillis) {
+                        interval = curInnerInterval * curRoundingInfo.roughEstimateDurationMillis;
+                        break roundingInfosLoop;
+                    }
+                }
+                roundingIdx++;
+            } while (
+                roundingIdx < roundingInfos.length
+            );
+
+            // interval > 0
+            int i = 0;
+            filters = new Weight[targetBuckets];
+            long roundedLow = preparedRounding.round(low);
+            while (i < targetBuckets) {
+                // Calculate the lower bucket bound
+                final byte[] lower = new byte[8];
+                NumericUtils.longToSortableBytes(Math.max(roundedLow, low), lower, 0);
+                // Calculate the upper bucket bound
+                final byte[] upper = new byte[8];
+                roundedLow = preparedRounding.round(roundedLow + interval);
+                // Subtract -1 if the minimum is roundedLow as roundedLow itself
+                // is included in the next bucket
+                NumericUtils.longToSortableBytes(Math.min(roundedLow - 1, high), upper, 0);
+
+                filters[i++] = context.searcher().createWeight(new PointRangeQuery(field, lower, upper, 1) {
+                    @Override
+                    protected String toString(int dimension, byte[] value) {
+                        return null;
+                    }
+                }, ScoreMode.COMPLETE_NO_SCORES, 1);
+            }
+        }
+
+        boolean tryFastFilterAggregation(LeafReaderContext ctx, long owningBucketOrd) throws IOException {
+            final int[] counts = new int[filters.length];
+            int i;
+            for (i = 0; i < filters.length; i++) {
+                counts[i] = filters[i].count(ctx);
+                if (counts[i] == -1) {
+                    // Cannot use the optimization if any of the counts
+                    // is -1 indicating the segment might have deleted documents
+                    return false;
+                }
+            }
+
+            for (i = 0; i < filters.length; i++) {
+                long bucketOrd = bucketOrds.add(
+                    owningBucketOrd,
+                    preparedRounding.round(NumericUtils.sortableBytesToLong(((PointRangeQuery) filters[i].getQuery()).getLowerPoint(), 0))
+                );
+                if (bucketOrd < 0) { // already seen
+                    bucketOrd = -1 - bucketOrd;
+                }
+                incrementBucketDocCount(bucketOrd, counts[i]);
+            }
+            throw new CollectionTerminatedException();
```

Comment:
We can be aggressive during bucket creation, and invoke the logic to merge buckets if it exceeds the target bucket count:

```
do {
                        try (LongKeyedBucketOrds oldOrds = bucketOrds) {
                            preparedRounding = prepareRounding(++roundingIdx);
                            long[] mergeMap = new long[Math.toIntExact(oldOrds.size())];
                            bucketOrds = new LongKeyedBucketOrds.FromSingle(context.bigArrays());
                            LongKeyedBucketOrds.BucketOrdsEnum ordsEnum = oldOrds.ordsEnum(0);
                            while (ordsEnum.next()) {
                                long oldKey = ordsEnum.value();
                                long newKey = preparedRounding.round(oldKey);
                                long newBucketOrd = bucketOrds.add(0, newKey);
                                mergeMap[(int) ordsEnum.ord()] = newBucketOrd >= 0 ? newBucketOrd : -1 - newBucketOrd;
                            }
                            merge(mergeMap, bucketOrds.size());
                        }
                    } while (roundingIdx < roundingInfos.length - 1
                        && (bucketOrds.size() > targetBuckets * roundingInfos[roundingIdx].getMaximumInnerInterval()
                            || max - min > targetBuckets * roundingInfos[roundingIdx].getMaximumRoughEstimateDurationMillis()));
```